http-server-netty/src/main/java/io/micronaut/http/server/netty/NettyHttpServer.java

```diff
@@ -657,4 +604,221 @@ public ChannelGroup getChannelGroup() {
+                .validateHttpHeaders(true)
+                .maxContentLength((int) serverConfiguration.getMaxRequestSize())
+                .build();
+
+        final HttpToHttp2ConnectionHandlerBuilder builder = new HttpToHttp2ConnectionHandlerBuilder()
+                .frameListener(http2ToHttpAdapter);
+
+        serverConfiguration.getLogLevel().ifPresent(logLevel ->
+                builder.frameLogger(new Http2FrameLogger(logLevel, NettyHttpServer.class))
+        );
+        return builder.connection(connection).build();
+    }
+
+    /**
+     * Negotiates with the browser if HTTP2 or HTTP is going to be used. Once decided, the Netty
+     * pipeline is setup with the correct handlers for the selected protocol.
+     */
+    private final class Http2OrHttpHandler extends ApplicationProtocolNegotiationHandler {
+
+        private final SslContext sslContext;
+
+        /**
+         * Default constructor.
+         * @param sslContext The SSL context
+         * @param fallbackProtocol The fallback protocol
+         */
+        Http2OrHttpHandler(SslContext sslContext, String fallbackProtocol) {
+            super(fallbackProtocol);
+            this.sslContext = sslContext;
+        }
+
+        @Override
+        public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+            if (evt instanceof SslHandshakeCompletionEvent) {
+                SslHandshakeCompletionEvent event = (SslHandshakeCompletionEvent) evt;
+                if (!event.isSuccess()) {
+                    final Throwable cause = event.cause();
+                    if (!(cause instanceof ClosedChannelException)) {
+                        super.userEventTriggered(ctx, evt);
+                    } else {
+                        return;
+                    }
+                }
+            }
+            super.userEventTriggered(ctx, evt);
+        }
+
+        @Override
+        protected void configurePipeline(ChannelHandlerContext ctx, String protocol) {
+            final ChannelPipeline pipeline = ctx.pipeline();
+            configurePipeline(protocol, pipeline);
+        }
+
+        private void configurePipeline(String protocol, ChannelPipeline pipeline) {
+            Map<String, ChannelHandler> handlers = getHandlerForProtocol(protocol);
+            handlers.forEach(pipeline::addLast);
+        }
+
+        @NotNull
+        private Map<String, ChannelHandler> getHandlerForProtocol(@Nullable String protocol) {
+            final HttpRequestDecoder requestDecoder = new HttpRequestDecoder(NettyHttpServer.this, environment, serverConfiguration);
+            final HttpRequestCertificateHandler requestCertificateHandler = new HttpRequestCertificateHandler();
+            final HttpResponseEncoder responseDecoder = new HttpResponseEncoder(mediaTypeCodecRegistry, serverConfiguration);
+            final RoutingInBoundHandler routingHandler = new RoutingInBoundHandler(
```

Comment:
`HttpRequestCertificateHandler` and `RoutingInBoundHandler` are `@Sharable`. Same instance can be used in all channels.

Reply:
Will look at this thanks